File: csharp/test/Ice/executor/Executor.cs
// Copyright (c) ZeroC, Inc.

using System.Diagnostics;

public class Executor
{
    private static void test(bool b) => global::Test.TestHelper.test(b);

    public Executor()
    {
        Debug.Assert(_instance == null);
        _instance = this;
        _thread = new Thread(run);
        _thread.Start();
    }

    public void run()
    {
        while (true)
        {
            System.Action call = null;
            lock (_m)
            {
                if (!_terminated && _calls.Count == 0)
                {
                    Monitor.Wait(_m);
                }

                if (_calls.Count > 0)
                {
                    call = _calls.Dequeue();
                }
                else if (_terminated)
                {
                    // Terminate only once all calls are dispatched.
                    return;
                }
            }

            if (call != null)
            {
                try
                {
                    call();
                }
                catch (System.Exception)
                {
                    // Exceptions should never propagate here.
                    test(false);
                }
            }
        }
    }

    public void execute(System.Action call, Ice.Connection con)
    {
        lock (_m)
        {
            _calls.Enqueue(call);
            if (_calls.Count == 1)
            {
                Monitor.Pulse(_m);
            }
        }
    }

    public static void terminate()
    {
        lock (_m)
        {
            _instance._terminated = true;
            Monitor.Pulse(_m);
        }

        _instance._thread.Join();
    }

    public static bool isExecutorThread() => Thread.CurrentThread == _instance._thread;

    private static Executor _instance;

    private readonly Queue<System.Action> _calls = new Queue<System.Action>();
    private readonly Thread _thread;
    private bool _terminated = false;
    private static readonly object _m = new object();
}

File: cpp/test/Ice/executor/Executor.cpp
// Copyright (c) ZeroC, Inc.

#include "Executor.h"
#include "TestHelper.h"

using namespace std;

shared_ptr<Executor> instance;
std::thread executorThread;

shared_ptr<Executor>
Executor::create()
{
    auto executor = shared_ptr<Executor>(new Executor());
    executorThread = std::thread([executor] { executor->run(); });
    instance = executor;
    return executor;
}

Executor::Executor() { _terminated = false; }

void
Executor::terminate()
{
    std::thread thread;
    {
        lock_guard lock(_mutex);
        _terminated = true;
        _conditionVariable.notify_one();
        thread = std::move(executorThread);
    }

    assert(thread.joinable());
    thread.join();
    instance = nullptr;
}

bool
Executor::isExecutorThread()
{
    return this_thread::get_id() == executorThread.get_id();
}

void
Executor::execute(const shared_ptr<ExecutorCall>& call, const shared_ptr<Ice::Connection>&)
{
    lock_guard lock(_mutex);
    _calls.push_back(call);
    if (_calls.size() == 1)
    {
        _conditionVariable.notify_one();
    }
}

void
Executor::run()
{
    while (true)
    {
        shared_ptr<ExecutorCall> call;
        {
            unique_lock lock(_mutex);
            _conditionVariable.wait(lock, [this] { return _terminated || !_calls.empty(); });

            if (!_calls.empty())
            {
                call = _calls.front();
                _calls.pop_front();
            }
            else if (_terminated)
            {
                // Terminate only once all calls are dispatched.
                return;
            }
        }

        if (call)
        {
            try
            {
                call->run();
            }
            catch (...)
            {
                // Exceptions should never propagate here.
                test(false);
            }
        }
    }
}
